map: Add SNAT fix option for specific ISP and displaying portsets in luci

missing233 opened the pull request:
1. Ensure that the multiple port ranges used with specific ISP MAP-E implementation are actually SNAT-ed correctly. 
from: https://github.com/fakemanhk/openwrt-jp-ipoe

2.add support for displaying portsets in luci.
![2](https://github.com/user-attachments/assets/2b4f8533-a043-4e93-91c9-fce3cf331a9c)

Dependent PR: https://github.com/openwrt/luci/pull/7301

missing233 committed:
map: Add SNAT fix option for specific ISP

Ensure that the multiple port ranges used with specific ISP
MAP-E implementation are actually SNAT-ed correctly.
from: https://github.com/fakemanhk/openwrt-jp-ipoe

Signed-off-by: Arayuki Mago <ms@missing233.com>

missing233 commented on package/network/ipv6/map/files/map.sh:
```diff
@@ -140,19 +139,51 @@ proto_map_setup() {
-	        json_close_object
-              done
-	    done
+	    if [ "$isp_fix" = "1" ]; then
```
Replaced with `snat_fix`.

missing233 commented:
@hauke I replaced it with `snat_fix`. How does that sound?

musashino205 commented:
Please fold the commit description within 74 characters.

https://openwrt.org/submitting-patches#submission_guidelines

> - commit description
>   - must have less than 75 characters per line

missing233 committed:
map: Add support for displaying portsets in luci

add support for showing MAP-E portsets on luci status page.
add fw3 support for snat fix

Signed-off-by: Arayuki Mago <ms@missing233.com>

Headcrabed commented:
Hello, anyone could give this MR a review?

Headcrabed commented:
@hauke @musashino205 Hello, can you give this MR a review? It is very useful.

Headcrabed commented:
@hauke  @musashino205  Seems the original maintainer of map (@dedeckeh ) seems is not maintaining anything for a long time(maybe over a year), so we can't wait for the original maintainer to merge this PR.

systemcrash commented:
ping

Headcrabed commented:
@systemcrash what is still needed before merging this path? I pinged multiple times but no reply.

systemcrash commented:
@jow-

missing233 commented:
Hi, just wanted to check in
is there anything else I can do?

systemcrash commented:
Things are pretty slow in the main repo. Just gotta wait a year.

missing233 commented:
@981213 
Hi, could you please help review my PR at your convenience? Thank you!

systemcrash commented on package/network/ipv6/map/files/map.sh:
```diff
@@ -140,19 +141,61 @@ proto_map_setup() {
+				local startport=$(echo $portset | cut -d'-' -f1)
+				local endport=$(echo $portset | cut -d'-' -f2)
+				for x in $(seq $startport $endport); do
+					if ! echo "$DONT_SNAT_TO" | tr ' ' '\n' | grep -qw $x; then
```
This part of the loop will be ... slow. Recommend optimising DONT_SNAT_TO to an array. You're making the new variable anyway, so you have the option to use an array in uci. This avoids grep tr and echo calls on every loop iteration.

systemcrash commented on package/network/ipv6/map/files/map.sh:
```diff
@@ -140,19 +141,61 @@ proto_map_setup() {
+			local portcount=0
+			local allports=""
+			for portset in $(eval "echo \$RULE_${k}_PORTSETS"); do
+				local startport=$(echo $portset | cut -d'-' -f1)
+				local endport=$(echo $portset | cut -d'-' -f2)
```
```sh
local startport=${portset%-*}
local endport=${portset#*-}
```

systemcrash commented on package/network/ipv6/map/files/map.sh:
```diff
@@ -101,7 +103,6 @@ proto_map_setup() {
 			fi
 		json_close_object
 
-
```
drop this

systemcrash commented on package/network/ipv6/map/files/map.sh:
```diff
@@ -140,19 +141,61 @@ proto_map_setup() {
+				iptables -t nat -N MAPE_CHAIN
+				iptables -t nat -A POSTROUTING -o "map-$cfg" -j MAPE_CHAIN
+				for proto in icmp tcp udp; do
+        			for port in $(echo "$allports" | sed 's/.*: //g' | tr ',' '\n' | tr -d ' '); do
```
Do we really have to do this? Can't we store a clean array for allports and just loop through it?

missing233 replied:
Changed to another implementation, but I'm not sure if it works properly.

systemcrash commented:
Maybe also ping three star shell general @pprindeville.

missing233 commented:
> Maybe also ping three star shell general @pprindeville.

I did my best to give it a try. But since this PR is from a community patch, and I’m not exactly a pro at writing shell scripts—and I don't have a MAP-E environment available for testing right now—I can't be sure if this code actually works as expected.

Neustradamus commented:
Hope that one another guy can review this PR :)

jow- commented:
Couple of questions:
- What is `fix_snat` fixing?
- Why can't it made be the default / only behavior? Why is the opt-in needed, is there any use for the presumably broken behavior?
- Direct nft / fw4 invocation in a proto handler is an absolute no-go, this needs to be abstracted away. Ideally in the form of a firewall include if the required rules cannot be expressed as proper uci rules

Neustradamus commented:
Important to specify that it is a support for the RFC7597: Mapping of Address and Port with Encapsulation (MAP-E):
- https://datatracker.ietf.org/doc/html/rfc7597

missing233 commented:
> Important to specify that it is a support for the RFC7597: Mapping of Address and Port with Encapsulation (MAP-E):
> 
> * https://datatracker.ietf.org/doc/html/rfc7597

this is actually a fix for bug in the draft of MAP-E used by Japan's NTT: 
https://datatracker.ietf.org/doc/html/draft-ietf-softwire-map-03#page-6

missing233 commented:
> Couple of questions:
> 
> * What is `fix_snat` fixing?
> * Why can't it made be the default / only behavior? Why is the opt-in needed, is there any use for the current, presumably broken behavior?
> * Direct nft / fw4 invocation in a proto handler is an absolute no-go, this needs to be abstracted away. Ideally in the form of a firewall include if the required rules cannot be expressed as proper uci rules

Hi, It seems like this feature can't be implemented in a more generic way. Does OpenWrt offer any more specific wrapper or implementation for nftables and iptables?

systemcrash commented:
Firewall basically talks with nft. Perhaps there are some utils available in the firewall?

missing233 commented:
It seems that this issue has already been resolved in OpenWrt 24.10.
